### PR TITLE
Fix orientations + XYZ slicing

### PR DESCRIPTION
**Pull request by agirault**
- [x] Use proper orientation matrix in ITKHelper
  - Needed to transpose it from ITK to VTK.
  - Tested with numerous datasets with orientation matrices different than identity.
- [x] Add X,Y,Z slicing (and rename previous IJK slicing for data space slicing)
  - Only supports it if X, Y, Z are orthogonal to I, J, K axis for now (aka flips and permutations). Will still do it if one of those component is > 0.99 (so with a small tilt). All world axis that do not perfectly align one of IJK axis should require a reformat to compute the slice, but this will be added in future work (@martinken promised :clinking_glasses: )
  - Still use indexes for X, Y, Z for now (instead of world coordinates) since they currently still map to IJK. In the future, X, Y, Z or any world axis slicing (ex: normal to the camera or to a plane widget...) should be in world coordinates (based on imagedata spacing/bounds).

### CODE REVIEW

**Comment by floryst** [comment]
LGTM

**Comment by agirault** [comment]
I'll be working on updating the image mapper to properly do X,Y,Z slicing. Could merge those together or separately.

**Comment by thewtex** [comment]
All wee needed was a


You get a 10.0, Alexis.

**Comment by agirault** [comment]
Almost ready!
- @jourdain: there are some API changes here (removing xSlice, ySlice, zSlice from mapper and slicerep). Enough to be considered breaking changes? Knowing that, when we add reformatting, we'll need to update the API so that X, Y, Z is not index based but world based, so that will change again.
- @martinken: currently failing tests with textures:
  - [ ]  vtkOpenGLImageMapper testImage not ok 413
  - [ ]  Image match resolution not ok 415 
  - [ ] vtkOpenGLImageMapper testImage not ok 417
  - [ ] mage match resolution not ok 419

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 1]
```diff
@@ -570,7 +575,7 @@ function vtkOpenGLImageMapper(publicAPI, model) {
         ptsArray[9] = ext[1];
         ptsArray[10] = nSlice;
         ptsArray[11] = ext[5];
-      } else {
+      } else if (ijkMode === SlicingMode.K) {
```
Comment (agirault) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: @martinken this seems to be the root of the tests failing: should this code run if we are not doing slicing? (`currentSlicingMode = SlicingMode.None`), and not just if we are doing K slicing? I thought this was only for what we used to call Z slicing.

**Comment by agirault** [comment]
Tests fixed, I was thinking of switching to using world coordinates and not indexed anymore for X, Y, Z.

**Comment by agirault** [comment]
tests and example pass, with X, Y, Z slicing using world coordinates and not indexed coordinates!
Ready for review @jourdain @martinken @floryst. Thank you :pray:

File: Sources/Proxy/Representations/SliceRepresentationProxy/index.js [hunk 2]
```diff
@@ -20,14 +20,14 @@ function mean(...array) {
...
   const axisIndex = 'XYZ'.indexOf(slicingMode);
 
   const propToUpdate = {
     sliceIndex: {
       domain: {
-        min: extent[axisIndex * 2],
-        max: extent[axisIndex * 2 + 1],
+        min: bounds[axisIndex * 2],
+        max: bounds[axisIndex * 2 + 1],
         step: 1,
```
Comment (jourdain) [inline Sources/Proxy/Representations/SliceRepresentationProxy/index.js]: Should this be updated as well to match the correct axis spacing?

Comment (agirault) [reply Sources/Proxy/Representations/SliceRepresentationProxy/index.js]: done

**Comment by jourdain** [comment]
Looks good enough for me... But what about you @martinken ?

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 3]
```diff
@@ -16,80 +17,191 @@ function vtkImageMapper(publicAPI, model) {
...
 
-    if (changeDetected && mtime === publicAPI.getMTime()) {
-      publicAPI.modified();
+    // Project vec3 onto direction cosines
+    const out = [0, 0, 0];
+    const a = publicAPI.getInputData().getDirection();
+    const mat3 = [[a[0], a[1], a[2]], [a[3], a[4], a[5]], [a[6], a[7], a[8]]];
+    const mat3I = [[0, 0, 0], [0, 0, 0], [0, 0, 0]];
+    vtkMath.invertMatrix(mat3, mat3I, 3);
+    vtkMath.multiply3x3_vect3(mat3I, inVec3, out);
```
Comment (martinken) [inline Sources/Rendering/Core/ImageMapper/index.js]: Maybe just reorder indicies as rotation inverse is just a transpose I think

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 4]
```diff
@@ -101,38 +213,80 @@ function vtkImageMapper(publicAPI, model) {
...
+      case SlicingMode.Y:
+        nSlice = publicAPI.getSliceAtPosition(model.slice);
+        extent[2] = nSlice - thickness;
+        extent[3] = nSlice + thickness;
+        break;
+      case SlicingMode.Z:
+        nSlice = publicAPI.getSliceAtPosition(model.slice);
+        extent[4] = nSlice - thickness;
+        extent[5] = nSlice + thickness;
+        break;
```
Comment (martinken) [inline Sources/Rendering/Core/ImageMapper/index.js]: Do not know that X==i here etc

File: Sources/Rendering/Core/ImageMapper/index.js (hunk 3 above)
Comment (agirault) [reply Sources/Rendering/Core/ImageMapper/index.js]: https://en.wikipedia.org/wiki/Rotation_matrix#Multiplication :raised_hands:

Comment (agirault) [reply Sources/Rendering/Core/ImageMapper/index.js]: Done: https://github.com/Kitware/vtk-js/pull/591/commits/cbdae0dd024f42e5bd75a4b9316d092a8bd38cf0#diff-9762803aa1d1e7c8613c71d155ed4e0dR89

File: Sources/Rendering/Core/ImageMapper/index.js (hunk 4 above)
Comment (agirault) [reply Sources/Rendering/Core/ImageMapper/index.js]: Done:
- https://github.com/Kitware/vtk-js/pull/591/commits/8b80c6ff33c104f6be3c450f7cad186a26afc26b#diff-9762803aa1d1e7c8613c71d155ed4e0dR215
- https://github.com/Kitware/vtk-js/pull/591/commits/8b80c6ff33c104f6be3c450f7cad186a26afc26b#diff-9762803aa1d1e7c8613c71d155ed4e0dR247

**Comment by agirault** [comment]
Last changes discussed with @martinken made. Also renamed `currentSlicingMode` to `slicingMode` and updated `croppingregionswidget` example to use IJK. I believe it's ready @jourdain !

**Comment by agirault** [comment]
@floryst could you give a look at the last commit? it updates the imageCroppingRegionsWidget example to use IJK.

**Comment by floryst** [comment]
LGTM

**Pull request merged**